src/metabase/lib/segment.cljc

```diff
@@ -67,21 +67,22 @@
    (available-segments query -1))
   ([query :- ::lib.schema/query
     stage-number :- :int]
-   (when-let [source-table-id (lib.util/source-table-id query)]
-     (let [segments (lib.metadata.protocols/segments (lib.metadata/->metadata-provider query) source-table-id)
-           segment-filters (into {}
-                                 (keep-indexed (fn [index filter-clause]
-                                                 (when (lib.util/clause-of-type? filter-clause :segment)
-                                                   [(get filter-clause 2) index])))
-                                 (lib.filter/filters query stage-number))]
-       (cond
-         (empty? segments)        nil
-         (empty? segment-filters) (vec segments)
-         :else                    (mapv (fn [segment-metadata]
-                                          (let [filter-pos (-> segment-metadata :id segment-filters)]
-                                            (cond-> segment-metadata
-                                              ;; even though at most one filter can reference a given segment
-                                              ;; we use plural in order to keep the interface used with
-                                              ;; plain filters referencing columns
-                                              filter-pos (assoc :filter-positions [filter-pos]))))
-                                        segments))))))
+   (when (zero? (lib.util/canonical-stage-index query stage-number))
```

Comment:
This seems like it should be marked as a temporary fix. 

If the first stage returns columns that would make a segment still valid in the second stage, it could be marked as available still, correct? Until we put that logic in place it makes sense to disallow but I'd like to see comments indicating such.

Reply:
Good point, there was a [discussion](https://metaboat.slack.com/archives/C01LQQ2UW03/p1701277003243209) about it, and this is the outcome.